### PR TITLE
Adds a new auto-interval date histogram

### LINKED ISSUE #9572: Support dynamic interval and fixed buckets for histogram aggregation
This is basically the opposite of how the histogram currently acts: fixed interval and dynamic number of buckets.

My use case is that I would like to build a price range style aggregation for filtering products:

$0 - $100
$101 - $200
$201 - $300
$301 - $400
$401 - $500

While you can do this with the range agg. if you know in advance the lower and upper bounds of all your prices, I can't see a way to do this dynamically based off the current query. The range agg also suffers from other shortcomings such as generating buckets that may contain no documents.

Ideally, I'd like to be able to specify the number of ranges / buckets I want (eg. 5) with a dynamic interval calculated by the aggregation.

`min_doc_count` could also come into affect here to ensure that either we don't generate buckets that don't contain any products (eg `min_doc_count: 0`) and in which case the number of buckets just becomes a maximum number of potential buckets, rather than a guaranteed amount.

### PR DESCRIPTION
**Pull request by colings86**
This change adds a new type of histogram aggregation called `auto_date_histogram` where you can specify the target number of buckets you require and it will find an appropriate interval for the returned buckets. The aggregation works by first collecting documents in buckets at second interval, when it has created more than the target number of buckets it merges these buckets into minute interval bucket and continues collecting until it reaches the target number of buckets again. It will keep merging buckets when it exceeds the target until either collection is finished or the highest interval (currently years) is reached. A similar process happens at reduce time.

This aggregation intentionally does not support min_doc_count, offest and extended_bounds to keep the already complex logic from becoming more complex. The aggregation accepts sub-aggregations but will always operate in `breadth_first` mode deferring the computation of sub-aggregations until the final buckets from the shard are known. min_doc_count is effectively hard-coded to zero meaning that we will insert empty buckets where necessary.

This is also the first aggregation to merge buckets at collection time but having this aggregation will open the door to other such aggregations that merge buckets at collection time.

Things still to do:
* [x] Write documentation on the new aggregation and its parameters
* [x] Write more tests, specifically around having sub-aggregations and checking they are merged correctly in the reduce step
* [x] Investigate whether we could collect `10 * target` buckets on the shards and then use the extra bucket to create buckets representing multiples of the interval in the cases where we have too many buckets for one interval but if we increase the interval we get only 1 (or a few) bucket(s) (e.g. if the target buckets is 10 and we end up with 30 x minute level buckets if we increase the rounding we'll get 1 x hour bucket so instead can we merge every 3 minute level buckets to get 10 x 3 minute buckets?) - this will probably be moved out into a separate issue to be tackled when this first pass is merged
* [x] Add an equivalent `auto_histogram` aggregation to work on numeric (rather than date) fields - this should be moved into a separate issue
* [x] Fail the request at parsing time if the number of buckets requested is at higher than soft_limit_value / m where m is the largest innerInterval for all rounding intervals (the largest number of buckets we would merge to one bucket in any rounding, currently 30)
* [x] Add a test to ensure that if a user requests more than 10k buckets, we return an error.
* [x] Fix doc tests

Closes #9572

### CODE REVIEW

**Comment by colings86** [comment]
@jpountz there are tests that need to be added to this PR and @pcsanwald is looking into benchmarking the aggregation at the moment to make sure the time performance is ok but I think its worth you taking a look so if there are any major things to change we can fix them now

**Review by jpountz — COMMENTED** [review]
I had a quick look and think the general approach is good. It would help to better document how the target number of buckets is interpreted and how far the actual number of buckets may be.

**Comment by colings86** [comment]
@jpountz I've pushed the tests that I mentioned were missing before and added the information you suggested to the documentation, I think this is ready for a full review now

**Comment by pcsanwald** [comment]
I've been doing some testing on this branch using a subset of the nyc_taxis dataset that rally uses. I've run into an an error that I don't think I should be seeing:
_query.json_
```
{
    "aggs" : {
        "dropoffs_over_time" : {
            "auto_date_histogram" : {
                "field" : "dropoff_datetime",
                "buckets" : 5000
            }
        }
    }
}
```

and then this query:
`curl -XPOST localhost:9200/_search -d @query.json`

gives me a max buckets error:

```
{
	"error": {
		"root_cause": [{
			"type": "too_many_buckets_exception",
			"reason": "Trying to create too many buckets. Must be less than or equal to: [10000] but was [133836]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
			"max_buckets": 10000
		}],
		"type": "search_phase_execution_exception",
		"reason": "all shards failed",
		"phase": "query",
		"grouped": true,
		"failed_shards": [{
			"shard": 0,
			"index": "nyc_taxis",
			"node": "3FtkNLQlQS-4KAFC6sP1Sw",
			"reason": {
				"type": "too_many_buckets_exception",
				"reason": "Trying to create too many buckets. Must be less than or equal to: [10000] but was [133836]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
				"max_buckets": 10000
			}
		}]
	},
	"status": 503
}
```

Is this a bug, or is this not a good test case? I'm happy to post a snapshot of the exact dataset I'm using someplace...

**Comment by colings86** [comment]
hmmm I think I know what is happening here and on the one hand the aggregation is doing the right thing by throwing an error but on the other hand I agree this is surprising to the user.

The crux of the problem is that although you asked for 5000 buckets in this case we need to create a lot more than that number of buckets to get the final answer. I am going to make some assumptions in this explanation so feel free to correct them if the data you have suggests something different is happening.

I think your index must be a single shard otherwise I would expect the request to be showing other shards to be failing too (since I would expect the data to be evenly and randomly distributed over all the shards). I also think that the shards are collecting at either second or minute level when this exception is thrown.

Let assume this is happening when collecting at second level. We will collect buckets with an interval of 1 second until we have more than `30 * target_buckets` buckets collected. At this point we know that we need to move up to the minute level as if we continued at the second level we would have more than `target_buckets` 30-second buckets in the final answer (we can only merge into the multiples of an interval as the very last step in the reduce phase). 

In your case you are asking for 5000 buckets and `30 * 5000 = 150000` which is way above the limit of 10000. The check on the number of buckets on the shards is done only when building the response to send back to the coordinating node so the aggregator happily collects the 133836 buckets but then trips this limit in `buildAggregation()` when we build the shard's response.

I'm not sure what we should do to fix this as we are definitely creating more buckets than the limit so the soft-limit is doing its job here. We could use the limit on the shard as another factor for when to merge up to the next level (i.e. have the threshold at `Math.min(LIMIT, 30 * target_buckets)` ) but that might also be surprising to the user as they could be returned far fewer buckets than they asked for and at a higher interval than is useful because the limit kept forcing the interval to be increased. 

Any ideas?

**Comment by pcsanwald** [comment]
@colings86 your analysis seems exactly right to me, I was indeed testing on a single shard. Unless we think of a better solution, I do think that returning far fewer buckets is a better user experience than the current behavior, although it's still not great and I'd love to come up with a solution that works within the soft limiting, I just can't think of one  😄

**Comment by polyfractal** [comment]
Perhaps a flag could be included in the response that lets the user know the buckets had to be merged due to the soft limit (and maybe how many times it had to do that, to give a feel for the magnitude that it was off by)?

And/or a flag that ran the agg in "strict" mode so the user would either get the requested number of buckets or fail when it hits the soft limit.

Not a great solution really, but perhaps better than nothing?

**Comment by pcsanwald** [comment]
@polyfractal I like both these suggestions. just brainstorming a bit, can we tell the max number of buckets that will not trip the limit in `buildAggregation()` for each interval? that could be useful info to return in an error

**Comment by colings86** [comment]
> can we tell the max number of buckets that will not trip the limit in `buildAggregation()` for each interval?

Yes we can, it would be `soft_limit_value / m` where `m` is the largest innerInterval for the current rounding interval (the largest number of buckets we would merge to one bucket in that rounding)

**Comment by colings86** [comment]
@jimczi and I had a chat about the bucket limit stuff for this aggregation, the below is a summary of what we discussed:

Firstly there are two sides to whats happening here. From the user thats sending the request this seems wrong because they asked for e.g. 5000 buckets and its claiming they have tripped at a 10000 bucket limit. From the cluster admins side this is exactly what they want to happen to protect the cluster, we are stopping a single request from consuming too many resources and destabilising the cluster. I think there is largely agreement that the limit is really doing the right thing for the cluster here and bypassing the limit would be a bad idea.

The problem with the approach I mentioned where we would collect `Math.min(LIMIT, 30 * target_buckets)` buckets before moving up to the next rounding interval is that this is still quite surprising for the user. They ask for e.g. 5000 buckets but they only get 300 because we changed the rules so we didn't trip the limit. This makes it hard for the user to reason about how many buckets they might get back from the aggregation.

@polyfractal's suggestion of adding information to the response and/or adding a strict mode is interesting but I wonder if we should just fail the request at parsing time if the number of buckets requested is at higher than `soft_limit_value / m` where m is the largest innerInterval for all rounding interval (the largest number of buckets we would merge to one bucket in any rounding, currently 30)? That way the user would be told before the request runs that its likely too heavy for the cluster to be able to service. Note that the current default bucket limit of 10000 allows the user to request 333 buckets without hitting the limit which already seems like a lot of data points to visualise. If users require more than this then the admin can increase the bucket limit but at least the admin gets the control of where the limit is and the user is able to reason about what the likely response is going to be.

One thing that complicates things a bit here is sub-aggregations. Regardless of how we deal with the limit in this aggregation, if it has sub-aggregations the limit could still be tripped anyway. I think thats ok though since this is no different from other aggregations.

Lastly, @jimczi had a suggestion that we could add an optional `rounding_hint` (naming TBD) which would tell the aggregation to start at a different rounding (the aggregation currently always starts at seconds) if the user knows that the aggregation is going to need at least e.g. day level buckets. This would help them potentially avoid the bucket limit and would also give a slight performance boost since we would perform a significantly smaller number of buckets whilst executing.

**Comment by pcsanwald** [comment]
@colings86 all this sounds good. I do wonder if there's a different way to express "number of buckets" that might serve us better and allow for more flexibility in the response? something like order of magnitude? I dunno, just an idea.

I do think that if we can change the error messaging to specify what the actual limit was at the time of the request, in a way that's programmatically accessible, would be helpful to a user (recognizing that the "actual limit" can and will change as new data is being indexed).

the case I'm thinking of is, if I'm building a visualization like a scatter plot where the plot point is Xpx and area is Ypx, I'm going to know roughly how many points will saturate the visualization because math. So if I request a max of 150 buckets and the limit turned out to be 120, I'd want to retry with that. This is just an example use case, but it is one from my experience building visualizations where I would have loved to have this autohisto feature.

**Comment by jpountz** [comment]
> I wonder if we should just fail the request at parsing time if the number of buckets requested is at higher than soft_limit_value / m

This is my preferred option. It doesn't require new options, fails early and should still be enough for the vast majority of users.

> Lastly, @jimczi had a suggestion that we could add an optional rounding_hint

I'd like to keep this idea for a follow-up iteration and make the API as minimal as possible for the first iteration.

**Comment by polyfractal** [comment]
>I wonder if we should just fail the request at parsing time if the number of buckets requested is at higher than soft_limit_value / m

I like this too.  I didn't realize earlier this was information you'd have upfront, I assumed you'd only find out while in the middle of merging buckets.

++ to what @pcsanwald said, if we can provide the "real" bucket value in errors that'd be helpful.  E.g. it'd confuse the user if we say "You can't execute this agg because it'd breach the soft-limit of 10000 buckets", when they only requested "150 auto buckets".

**Comment by colings86** [comment]
@pcsanwald I updated the description of this PR to include the outstanding tasks. It was mostly up to date already so I just tweaked it a bit. Let me know if you need any more detail on the points.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregatorTests.java [hunk 1]
```diff
@@ -0,0 +1,1312 @@
...
+                    assertEquals(2, bucket.getDocCount());
+
+                    bucket = buckets.get(2);
+                    assertEquals("2017-03-01T00:00:00.000Z", bucket.getKeyAsString());
+                    assertEquals(3, bucket.getDocCount());
+                }
+        );
+    }
+
+    public void testWithLargeNumberOfBuckets() {
```
Comment (pcsanwald) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregatorTests.java]: @colings86 is this what you had in mind for 10k buckets test? I couldn't find the constant for 10k anywhere else, which led me to think that this isn't the right place to enforce the limit

Comment (colings86) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregatorTests.java]: I commented on the builder about this as I wanted to reference the code there

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        roundings[3] = new RoundingInfo(createRounding(DateTimeUnit.DAY_OF_MONTH), 24 * 60 * 60 * 1000L, 1, 7);
+        roundings[4] = new RoundingInfo(createRounding(DateTimeUnit.MONTH_OF_YEAR), 30 * 24 * 60 * 60 * 1000L, 1, 3);
+        roundings[5] = new RoundingInfo(createRounding(DateTimeUnit.YEAR_OF_CENTURY), 365 * 24 * 60 * 60 * 1000L, 1, 5, 10, 20, 50, 100);
+
+        int maxRoundingInterval = Arrays.stream(roundings,0, roundings.length-1)
+            .map(rounding -> rounding.innerIntervals)
+            .flatMapToInt(Arrays::stream)
+            .boxed()
+            .reduce(Integer::max).get();
+        // TODO[PCS] what's the best way to get a hold of SearchService here, so I'm not just using default?
```
Comment (pcsanwald) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java]: @colings86 is this better in tems of implementation? the one outstanding thing with this (can chalk up to my lack of familiarity with guice and/or codebase in general) is how to get a hold of a `SearchService` instance here, which presumably would give me the correct maxbuckets setting

Comment (colings86) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java]: Yes this is what I meant for the implementation, thanks. In terms of getting the setting value @jpountz pointed out we can get hold of the node settings from `searchContext.getQueryShardContext().getIndexSettings().getNodeSettings()` and then we can pass that `Settings` object to `MultiBucketConsumerService.MAX_BUCKET_SETTING.get(Settings)` to get the actual value of the setting.

**Comment by jpountz** [comment]
This aggregation has a different set of options, it adds a new `buckets` option and doesnt't support `min_doc_count` and `offset`. Having a distinct aggregation is less confusing in my opinion compared to making documentation and validation logic more complex.

**Comment by pcsanwald** [comment]
@jpountz I've added the check in `innerBuild` for the calculation we agreed on, a test for bucket limit, and an appropriate error message. This is ready for another review when you have a chance.

**Comment by pcsanwald** [comment]
Opened #31828 as a separate issue for numeric aggregation.

**Comment by pcsanwald** [comment]
@colings86 this is ready for review when you have time.

**Review by colings86 — COMMENTED** [review]
@pcsanwald I think what is here is good, but @timroes pointed out that there is currently not an easy way for the user to know what interval we ended up using for the buckets. At the moment they would need to try to guess the interval by comparing two consecutive buckets which is not particularly nice. I wonder if we should add a field outside of the buckets in `InternalAutoDateHistogram` itself that shows the final interval in the response so the user can use this?

**Review by colings86 — COMMENTED** [review]
LGTM

**Comment by pcsanwald** [comment]
@colings86 and @polyfractal and I just discussed, we like @timroes suggestion and I will implement in a subsequent PR. this one is quite large and we'd prefer to review and merge and issue a smaller, more targeted PR to expose min interval.

**Comment by polyfractal** [comment]
Left a few small comments and questions, mostly superficial stuff.  Sorta glossed over the algo part, since others looked at that more closely.  I think it looks good!

**Comment by pcsanwald** [comment]
@polyfractal I believe I've addressed all your comments and this is ready for final review when you have a chance. note that I've tested locally and confirmed the bucket default is indeed 10 if not specified.

File: docs/reference/aggregations/bucket/autodatehistogram-aggregation.asciidoc [hunk 3]
```diff
@@ -6,6 +6,8 @@ instead of providing an interval to use as the width of each bucket, a target nu
 indicating the number of buckets needed and the interval of the buckets is automatically chosen to best achieve
 that target. The number of buckets returned will always be less than or equal to this target number.
 
+The buckets field is optional, and will default to 10 buckers if not specified.
```
Comment (polyfractal) [inline docs/reference/aggregations/bucket/autodatehistogram-aggregation.asciidoc]: Just a minor typo: `buckers` vs `buckets` :)

**Review by polyfractal — APPROVED** [review]
LGTM!

**Comment by mrec** [comment]
Apologies for the necropost, but I'm increasingly unsure that I've understood the implications of "*Fail the request at parsing time if the number of buckets requested is at higher than soft_limit_value / m*", mostly because I can't see where `soft_limit_value` is coming from. Is this a global ES config property? 

Also, the [agg docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-autodatehistogram-aggregation.html) don't make it clear what happens if the requested `buckets` can't be met even with 100-year intervals - I assume error again, although that feels a bit harsh since it's only partially under the caller's control - or whether the 7-day and 3-month roundings will be to natural (calendar) week/quarter boundaries.

**Comment by colings86** [comment]
@mrec The `soft_limit_value` refers to the existing soft limit that applies across all aggregations at limits the number of buckets that can be created by an aggregations request. You can see the setting defined in the code [here](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/MultiBucketConsumerService.java#L42).

The idea of the soft limits is that we have controls that aim to stop a single user performing a request that would destabilise or be otherwise detrimental to the cluster or would adversely affect other users trying to perform request on the cluster. Most of these soft limits end up effectively limiting the amount of memory a user's request can use in various ways but some do other things.

The idea with the "Fail the request at parsing time if the number of buckets requested is at higher than soft_limit_value / m" is that if at the point we parse the request we can already determine that we are going to need to produce more buckets than this particular soft limit allows there is not point in us executing the request since during the execution it will definitely trip the soft limit and fail. So instead we fail the request at parsing time.

Thats not to say that the soft limit won't be tripped if it passes the `soft_limit_value/m` test. Other factors like the number of buckets any sub aggregations create could still make us hit the soft limit but the idea is to catch the case where we already know at parse time that the request won't work.

If the request cannot be satisfied even with 100 year intervals then the request will fail on the soft limit as will an other aggregation request. I'd like to note two things here though:
1. As is explained in the error response upon hitting the soft limit and admin can modify the value of the soft limit if they wish. This is an admin choice (someone who can control cluster settings) because its something that needs to be considered with the health of the whole cluster and all its users in mind.
2. The default soft limit for the number of buckets is 10,000 so in order to trip the soft limit on the `soft_limit_value / m` test you would need to request `10,000 / 30 = 333` buckets in this aggregation (m is 30 because we take the maximum innerInterval of the all the intervals except the highest one) so assuming the data returned by the query would require 100-year buckets you are still able to aggregate a range of 33,000 years which to me seems very reasonable for a default.

**Comment by mrec** [comment]
@colings86 I think you're confirming my worries - a maximum range of 33,000 years is more than generous, but I think a (default) maximum `buckets` of 333 is too low. Particularly since it's not documented, it's not the case for other bucketed aggs and the following *"if a user requests more than 10k buckets, we return an error"* kind of implies that *that's* the limit.

**Comment by colings86** [comment]
> I think a (default) maximum buckets of 333 is too low

I don't agree here. Given that the final interval is out of the users hands I don't think this aggregation is useful for analytics jobs which are consuming a lot of buckets (since they would almost certainly want to determine the interval of the buckets themselves) but rather for visualising data to the user (e.g. on a bar chart). Given this I find it hard to believe that a visualisation could provide the information in a way that a user could consume if its rendering 100s of buckets. Is your use case intending to present the out put of `auto_date_histogram` to the user through a visualisation or to an analytics job?

> Particularly since it's not documented, it's not the case for other bucketed aggs and the following "if a user requests more than 10k buckets, we return an error" kind of implies that that's the limit.

We do document the soft limit [here](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/search-aggregations-bucket.html). Note that the documentation says its disabled by default at the moment but it will be set to a default of 10,000 in 7.0. The optimisation we added in this aggregation doesn't change the behaviour or lower the limit, it just catches the violation of the soft limit earlier before we start executing the request. As I mentioned before, we feel that the soft limits are important tools for cluster admins to be able to add confidence that a single request is not going to adversely affect the cluster or other user's requests and the limits can always be adjusted if the admin determines its appropriate to increase it, but then the decision can be made with the knowledge that the cluster is safe to handle these more expensive requests

**Comment by mrec** [comment]
Sorry, I think that came across more argumentative than intended. Our main expected use cases are around visualizations for screening, yes, and examples I've seen of "the sort of thing we want" include pixel-per-bar charts around 400px wide. Whether that resolution is actually *needed* is open to debate; I'll talk to our PM.

I do think it'd be worth calling out the 333 limit explicitly in the agg docs - the connection between that and the documented 10k limit is not obvious, and most users won't be reading this PR thread.

**Comment by colings86** [comment]
No offence taken at all 😄 

> I do think it'd be worth calling out the 333 limit explicitly in the agg docs - the connection between that and the documented 10k limit is not obvious, and most users won't be reading this PR thread.

You have convinced me on this. I think we should point this out in the documentation. Thanks for persevering with me

**Comment by colings86** [comment]
@mrec I raised https://github.com/elastic/elasticsearch/issues/32950 for the documentation issue

**Pull request merged**